src/main.rs

```diff
@@ -79,51 +71,121 @@ async fn transfer(
-            }));
-            // println!("packet: {:?}", packet)
+            tokio::spawn(async move { logger.write().await.handle_connect(p, &ip).await });
         }
-        io::copy(&mut ri, &mut wo).await?;
-        let r = wo.shutdown().await;
-        r
+
+        // copy packets from ri to wo
+        let packet_queue: VecDeque<(BytesMut, tokio::time::Instant)> = VecDeque::new();
+        let packet_queue = Arc::new(std::sync::Mutex::new(packet_queue));
+
+        let task_packet_queue = packet_queue.clone();
+
+        // read from the queue and write to write_outbound
+        let read_from_queue = async move {
+            loop {
+                // check if there's something in the packet queue every simulated_ping ms
+                tokio::time::sleep(StdDuration::from_millis(simulated_ping)).await;
+                // if there is, wait until the packet is ready to be sent and send it
+                loop {
+                    let queue_front = packet_queue.lock().unwrap().pop_front();
+                    if let Some((bytes, sent_at)) = &queue_front {
+                        let sending_at = *sent_at + StdDuration::from_millis(simulated_ping);
+                        tokio::time::sleep_until(sending_at).await;
+                        if write_outbound.write_all(bytes.as_ref()).await.is_err() {
+                            break;
+                        }
+                    } else {
+                        break;
+                    }
+                }
+            }
+        };
+
+        let write_to_queue = async move {
+            let mut framed = FramedRead::new(read_inbound, BytesCodec::new());
+            while let Some(message) = framed.next().await {
+                match message {
+                    Ok(bytes) => {
+                        task_packet_queue
+                            .lock()
+                            .unwrap()
+                            .push_back((bytes, tokio::time::Instant::now()));
+                    }
+                    Err(_) => break,
+                }
+            }
+        };
+
+        tokio::join!(read_from_queue, write_to_queue);
     };
 
-    
     let server_to_client = async {
-        io::copy(&mut ro, &mut wi).await?;
-        wi.shutdown().await
-    };
-    match tokio::try_join!(
-        tokio::time::timeout(StdDuration::from_secs(timeout), client_to_server),
-        tokio::time::timeout(StdDuration::from_secs(timeout), server_to_client)
-    ) {
-        Ok(_) => {}
-        Err(_) => {}
+        let mut outbound_framed = FramedRead::new(read_outbound, BytesCodec::new());
+        // copy packets from ro to wi
+        while let Some(message) = outbound_framed.next().await {
+            match message {
+                Ok(bytes) => {
+                    if write_inbound.write_all(&bytes).await.is_err() {
+                        break;
+                    }
+                }
+                Err(_) => break,
+            }
+        }
+
```

Comment:
is there a reason to do this over io::copy, as imo io::copy is more descriptive, plus it uses less ram.

Reply:
I'm pretty sure it uses the same amount of ram, I just made it like that so it's easier to add fake lag for server->client packets later.